File: testsuite/Jogl4Examples/ShadersExample/src/options/CommandLineOptions.java
package options;

import model.ShaderOperationMode;
import vsdk.toolkit.common.RendererConfiguration;

public class CommandLineOptions
{
    public enum TextureFilterOption {
        LINEAR,
        NEAREST
    }
    private static final int DEFAULT_OFFLINE_WIDTH = 1100;
    private static final int DEFAULT_OFFLINE_HEIGHT = 900;

    private boolean offline;
    private String offlineOutputPath;
    private ShaderOperationMode method;
    private Double rotationDegrees;
    private Double lightRotationDegrees;
    private Boolean withTexture;
    private Boolean withBumpMap;
    private Integer shadingType;
    private TextureFilterOption textureFilter;
    private Integer meridians;
    private Integer parallels;
    private Double cpuTextureOffsetUTexels;
    private Double cpuTextureOffsetVTexels;
    private boolean showHud;
    private int width;
    private int height;

    public static CommandLineOptions parse(String[] args)
    {
        CommandLineOptions options = new CommandLineOptions();
        options.offline = false;
        options.offlineOutputPath = null;
        options.method = ShaderOperationMode.OPENGL_4_1;
        options.rotationDegrees = null;
        options.lightRotationDegrees = null;
        options.withTexture = null;
        options.withBumpMap = null;
        options.shadingType = null;
        options.textureFilter = null;
        options.meridians = null;
        options.parallels = null;
        options.cpuTextureOffsetUTexels = null;
        options.cpuTextureOffsetVTexels = null;
        options.showHud = false;
        options.width = DEFAULT_OFFLINE_WIDTH;
        options.height = DEFAULT_OFFLINE_HEIGHT;

        for ( int i = 0; i < args.length; i++ ) {
            String arg = args[i];
            if ( "--offline".equals(arg) ) {
                ensureHasValue(args, i, "--offline");
                options.offline = true;
                options.offlineOutputPath = args[++i];
                continue;
            }
            if ( arg.startsWith("--offline=") ) {
                options.offline = true;
                options.offlineOutputPath = arg.substring("--offline=".length());
                continue;
            }
            if ( "--method".equals(arg) ) {
                ensureHasValue(args, i, "--method");
                options.method = parseMethod(args[++i]);
                continue;
            }
            if ( arg.startsWith("--method=") ) {
                options.method = parseMethod(arg.substring("--method=".length()));
                continue;
            }
            if ( "--rotation".equals(arg) ) {
                ensureHasValue(args, i, "--rotation");
                options.rotationDegrees = parseDouble(args[++i], "--rotation");
                continue;
            }
            if ( arg.startsWith("--rotation=") ) {
                options.rotationDegrees = parseDouble(
                    arg.substring("--rotation=".length()), "--rotation");
                continue;
            }
            if ( "--light-rotation".equals(arg) ) {
                ensureHasValue(args, i, "--light-rotation");
                options.lightRotationDegrees =
                    parseDouble(args[++i], "--light-rotation");
                continue;
            }
            if ( arg.startsWith("--light-rotation=") ) {
                options.lightRotationDegrees = parseDouble(
                    arg.substring("--light-rotation=".length()),
                    "--light-rotation");
                continue;
            }
            if ( "--with".equals(arg) ) {
                ensureHasValue(args, i, "--with");
                applyFeatureSwitches(options, args[++i], true);
                continue;
            }
            if ( arg.startsWith("--with=") ) {
                applyFeatureSwitches(
                    options,
                    arg.substring("--with=".length()),
                    true);
                continue;
            }
            if ( "--shading".equals(arg) ) {
                ensureHasValue(args, i, "--shading");
                options.shadingType = parseShading(args[++i]);
                continue;
            }
            if ( arg.startsWith("--shading=") ) {
                options.shadingType = parseShading(
                    arg.substring("--shading=".length()));
                continue;
            }
            if ( "--without".equals(arg) ) {
                ensureHasValue(args, i, "--without");
                applyFeatureSwitches(options, args[++i], false);
                continue;
            }
            if ( arg.startsWith("--without=") ) {
                applyFeatureSwitches(
                    options,
                    arg.substring("--without=".length()),
                    false);
                continue;
            }
            if ( "--texture-filter".equals(arg) ) {
                ensureHasValue(args, i, "--texture-filter");
                options.textureFilter = parseTextureFilter(args[++i]);
                continue;
            }
            if ( arg.startsWith("--texture-filter=") ) {
                options.textureFilter = parseTextureFilter(
                    arg.substring("--texture-filter=".length()));
                continue;
            }
            if ( "--meridians".equals(arg) ) {
                ensureHasValue(args, i, "--meridians");
                options.meridians = Integer.valueOf(Math.max(3, parseInt(args[++i], "--meridians")));
                continue;
            }
            if ( arg.startsWith("--meridians=") ) {
                options.meridians = Integer.valueOf(Math.max(3, parseInt(
                    arg.substring("--meridians=".length()), "--meridians")));
                continue;
            }
            if ( "--parallels".equals(arg) ) {
                ensureHasValue(args, i, "--parallels");
                options.parallels = Integer.valueOf(Math.max(2, parseInt(args[++i], "--parallels")));
                continue;
            }
            if ( arg.startsWith("--parallels=") ) {
                options.parallels = Integer.valueOf(Math.max(2, parseInt(
                    arg.substring("--parallels=".length()), "--parallels")));
                continue;
            }
            if ( "--cpu-texture-offset-u".equals(arg) ) {
                ensureHasValue(args, i, "--cpu-texture-offset-u");
                options.cpuTextureOffsetUTexels =
                    Double.valueOf(parseDouble(args[++i], "--cpu-texture-offset-u"));
                continue;
            }
            if ( arg.startsWith("--cpu-texture-offset-u=") ) {
                options.cpuTextureOffsetUTexels = Double.valueOf(parseDouble(
                    arg.substring("--cpu-texture-offset-u=".length()),
                    "--cpu-texture-offset-u"));
                continue;
            }
            if ( "--cpu-texture-offset-v".equals(arg) ) {
                ensureHasValue(args, i, "--cpu-texture-offset-v");
                options.cpuTextureOffsetVTexels =
                    Double.valueOf(parseDouble(args[++i], "--cpu-texture-offset-v"));
                continue;
            }
            if ( arg.startsWith("--cpu-texture-offset-v=") ) {
                options.cpuTextureOffsetVTexels = Double.valueOf(parseDouble(
                    arg.substring("--cpu-texture-offset-v=".length()),
                    "--cpu-texture-offset-v"));
                continue;
            }
            if ( "--width".equals(arg) ) {
                ensureHasValue(args, i, "--width");
                options.width = Math.max(1, parseInt(args[++i], "--width"));
                continue;
            }
            if ( arg.startsWith("--width=") ) {
                options.width = Math.max(1, parseInt(
                    arg.substring("--width=".length()), "--width"));
                continue;
            }
            if ( "--height".equals(arg) ) {
                ensureHasValue(args, i, "--height");
                options.height = Math.max(1, parseInt(args[++i], "--height"));
                continue;
            }
            if ( arg.startsWith("--height=") ) {
                options.height = Math.max(1, parseInt(
                    arg.substring("--height=".length()), "--height"));
                continue;
            }
            if ( "--hud".equals(arg) ) {
                ensureHasValue(args, i, "--hud");
                options.showHud = parseBooleanSwitch(args[++i], "--hud");
                continue;
            }
            if ( arg.startsWith("--hud=") ) {
                options.showHud = parseBooleanSwitch(
                    arg.substring("--hud=".length()),
                    "--hud");
                continue;
            }

            throw new IllegalArgumentException("Unknown option: " + arg);
        }

        if ( options.offline && (options.offlineOutputPath == null ||
             options.offlineOutputPath.isBlank()) ) {
            throw new IllegalArgumentException(
                "--offline requires an output file path");
        }

        return options;
    }

    private static void ensureHasValue(String[] args, int i, String name)
    {
        if ( i + 1 >= args.length ) {
            throw new IllegalArgumentException(name + " requires a value");
        }
    }

    private static int parseInt(String raw, String optionName)
    {
        try {
            return Integer.parseInt(raw);
        }
        catch ( NumberFormatException e ) {
            throw new IllegalArgumentException(
                "Invalid integer for " + optionName + ": " + raw, e);
        }
    }

    private static double parseDouble(String raw, String optionName)
    {
        try {
            return Double.parseDouble(raw);
        }
        catch ( NumberFormatException e ) {
            throw new IllegalArgumentException(
                "Invalid number for " + optionName + ": " + raw, e);
        }
    }

    private static boolean parseBooleanSwitch(String raw, String optionName)
    {
        String normalized = raw.trim().toLowerCase();
        if ( normalized.equals("on") || normalized.equals("true") ||
             normalized.equals("1") || normalized.equals("yes") ) {
            return true;
        }
        if ( normalized.equals("off") || normalized.equals("false") ||
             normalized.equals("0") || normalized.equals("no") ) {
            return false;
        }
        throw new IllegalArgumentException(
            "Unknown value for " + optionName + ": " + raw +
            ". Use on/off.");
    }

    private static ShaderOperationMode parseMethod(String raw)
    {
        String normalized = raw.trim().toLowerCase();
        if ( normalized.equals("opengl") || normalized.equals("opengl_4_1") ) {
            return ShaderOperationMode.OPENGL_4_1;
        }
        if ( normalized.equals("software") ) {
            return ShaderOperationMode.SOFTWARE;
        }
        throw new IllegalArgumentException(
            "Unknown --method value: " + raw + ". Use opengl or software.");
    }

    private static void applyFeatureSwitches(
        CommandLineOptions options,
        String rawList,
        boolean enabled)
    {
        String[] tokens = rawList.split(",");
        for ( String rawToken : tokens ) {
            String token = rawToken.trim().toLowerCase();
            if ( token.isEmpty() ) {
                continue;
            }
            if ( token.equals("texture") || token.equals("textures") ) {
                options.withTexture = Boolean.valueOf(enabled);
                continue;
            }
            if ( token.equals("bumpmap") || token.equals("bump") ||
                 token.equals("normalmap") ) {
                options.withBumpMap = Boolean.valueOf(enabled);
                continue;
            }
            throw new IllegalArgumentException(
                "Unknown feature in --with/--without: " + rawToken +
                ". Use texture,bumpmap");
        }
    }

    private static int parseShading(String raw)
    {
        String normalized = raw.trim().toLowerCase();
        if ( normalized.equals("constant") || normalized.equals("nolight") ) {
            return RendererConfiguration.SHADING_TYPE_NOLIGHT;
        }
        if ( normalized.equals("flat") ) {
            return RendererConfiguration.SHADING_TYPE_FLAT;
        }
        if ( normalized.equals("gouraud") ) {
            return RendererConfiguration.SHADING_TYPE_GOURAUD;
        }
        if ( normalized.equals("phong") ) {
            return RendererConfiguration.SHADING_TYPE_PHONG;
        }
        if ( normalized.equals("cook") ||
             normalized.equals("cook_torrance") ||
             normalized.equals("cook-torrance") ||
             normalized.equals("cooktorrance") ) {
            return RendererConfiguration.SHADING_TYPE_COOK_TERRANCE;
        }
        throw new IllegalArgumentException(
            "Unknown --shading value: " + raw +
            ". Use constant, flat, gouraud, phong, or cook_torrance.");
    }

    private static TextureFilterOption parseTextureFilter(String raw)
    {
        String normalized = raw.trim().toLowerCase();
        if ( normalized.equals("linear") ) {
            return TextureFilterOption.LINEAR;
        }
        if ( normalized.equals("nearest") ) {
            return TextureFilterOption.NEAREST;
        }
        throw new IllegalArgumentException(
            "Unknown --texture-filter value: " + raw +
            ". Use linear or nearest.");
    }

    public boolean isOffline()
    {
        return offline;
    }

    public String getOfflineOutputPath()
    {
        return offlineOutputPath;
    }

    public ShaderOperationMode getMethod()
    {
        return method;
    }

    public Double getRotationDegrees()
    {
        return rotationDegrees;
    }

    public int getWidth()
    {
        return width;
    }

    public int getHeight()
    {
        return height;
    }

    public Double getLightRotationDegrees()
    {
        return lightRotationDegrees;
    }

    public Boolean getWithTexture()
    {
        return withTexture;
    }

    public Boolean getWithBumpMap()
    {
        return withBumpMap;
    }

    public Integer getShadingType()
    {
        return shadingType;
    }

    public TextureFilterOption getTextureFilter()
    {
        return textureFilter;
    }

    public Integer getMeridians()
    {
        return meridians;
    }

    public Integer getParallels()
    {
        return parallels;
    }

    public Double getCpuTextureOffsetUTexels()
    {
        return cpuTextureOffsetUTexels;
    }

    public Double getCpuTextureOffsetVTexels()
    {
        return cpuTextureOffsetVTexels;
    }

    public boolean getShowHud()
    {
        return showHud;
    }
}

File: cpp/testsuite/OpenGL4Examples/ShadersExample/src/options/CommandLineOptions.cpp
#include "CommandLineOptions.h"

#include <algorithm>
#include <cctype>
#include <cstdlib>
#include <stdexcept>

#include "vsdk/toolkit/environment/material/RendererConfiguration.h"

CommandLineOptions::CommandLineOptions()
    : offline(false),
      method(ShaderOperationMode::OPENGL_4_1),
      hasRotation(false), rotationDegrees(0.0),
      hasLightRotation(false), lightRotationDegrees(0.0),
      hasWithTexture(false), withTexture(true),
      hasWithBumpMap(false), withBumpMap(true),
      hasShadingType(false), shadingType(RendererConfiguration::SHADING_TYPE_PHONG),
      hasTextureFilter(false), textureFilter(TextureFilterOption::LINEAR),
      hasMeridians(false), meridians(100),
      hasParallels(false), parallels(50),
      hasCpuTextureOffsetU(false), cpuTextureOffsetUTexels(-0.5),
      hasCpuTextureOffsetV(false), cpuTextureOffsetVTexels(-0.5),
      showHud(false), width(1100), height(900) {}

static java::String lower(const java::String& s)
{
    java::String out = s;
    std::transform(out.begin(), out.end(), out.begin(), [](unsigned char c){ return (char)std::tolower(c); });
    return out;
}

static bool startsWith(const java::String& s, const java::String& p)
{
    return s.size() >= p.size() && s.substr(0, p.size()) == p;
}

static ShaderOperationMode parseMethod(const java::String& raw)
{
    const java::String v = lower(raw);
    if (v == "opengl" || v == "opengl_4_1") return ShaderOperationMode::OPENGL_4_1;
    if (v == "software") return ShaderOperationMode::SOFTWARE;
    throw std::invalid_argument("Unknown --method value: " + raw + ". Use opengl or software.");
}

static int parseShading(const java::String& raw)
{
    const java::String v = lower(raw);
    if (v == "constant" || v == "nolight") return RendererConfiguration::SHADING_TYPE_NOLIGHT;
    if (v == "flat") return RendererConfiguration::SHADING_TYPE_FLAT;
    if (v == "gouraud") return RendererConfiguration::SHADING_TYPE_GOURAUD;
    if (v == "phong") return RendererConfiguration::SHADING_TYPE_PHONG;
    if (v == "cook" || v == "cook_torrance" || v == "cook-torrance" || v == "cooktorrance") {
        return RendererConfiguration::SHADING_TYPE_COOK_TERRANCE;
    }
    throw std::invalid_argument("Unknown --shading value: " + raw);
}

static CommandLineOptions::TextureFilterOption parseTextureFilter(const java::String& raw)
{
    const java::String v = lower(raw);
    if (v == "linear") return CommandLineOptions::TextureFilterOption::LINEAR;
    if (v == "nearest") return CommandLineOptions::TextureFilterOption::NEAREST;
    throw std::invalid_argument("Unknown --texture-filter value: " + raw);
}

static void applyFeatureSwitches(CommandLineOptions& o, const java::String& csv, bool enabled)
{
    size_t b = 0;
    while (b < csv.size()) {
        size_t e = csv.find(',', b);
        if (e == java::String::npos) e = csv.size();
        const java::String t0 = lower(csv.substr(b, e - b));
        java::String t;
        for (size_t i = 0; i < t0.size(); ++i) {
            const unsigned char c = (unsigned char)t0[i];
            if (!std::isspace(c)) {
                char oneChar[2] = { t0[i], '\0' };
                t += oneChar;
            }
        }
        if (t == "texture" || t == "textures") {
            o.hasWithTexture = true;
            o.withTexture = enabled;
        }
        else if (t == "bumpmap" || t == "bump" || t == "normalmap") {
            o.hasWithBumpMap = true;
            o.withBumpMap = enabled;
        }
        else if (!t.empty()) {
            throw std::invalid_argument("Unknown feature in --with/--without: " + t);
        }
        b = e + 1;
    }
}

CommandLineOptions CommandLineOptions::parse(int argc, char** argv)
{
    CommandLineOptions o;
    for (int i = 1; i < argc; i++) {
        java::String arg(argv[i]);
        auto readValue = [&](const java::String& name)->java::String {
            if (i + 1 >= argc) throw std::invalid_argument(name + " requires a value");
            return java::String(argv[++i]);
        };

        if (arg == "--offline") { o.offline = true; o.offlineOutputPath = readValue("--offline"); continue; }
        if (startsWith(arg, "--offline=")) { o.offline = true; o.offlineOutputPath = arg.substr(10); continue; }
        if (arg == "--method") { o.method = parseMethod(readValue("--method")); continue; }
        if (startsWith(arg, "--method=")) { o.method = parseMethod(arg.substr(9)); continue; }
        if (arg == "--rotation") { o.hasRotation = true; o.rotationDegrees = std::atof(readValue("--rotation").c_str()); continue; }
        if (startsWith(arg, "--rotation=")) { o.hasRotation = true; o.rotationDegrees = std::atof(arg.substr(11).c_str()); continue; }
        if (arg == "--light-rotation") { o.hasLightRotation = true; o.lightRotationDegrees = std::atof(readValue("--light-rotation").c_str()); continue; }
        if (startsWith(arg, "--light-rotation=")) { o.hasLightRotation = true; o.lightRotationDegrees = std::atof(arg.substr(17).c_str()); continue; }
        if (arg == "--with") { applyFeatureSwitches(o, readValue("--with"), true); continue; }
        if (startsWith(arg, "--with=")) { applyFeatureSwitches(o, arg.substr(7), true); continue; }
        if (arg == "--without") { applyFeatureSwitches(o, readValue("--without"), false); continue; }
        if (startsWith(arg, "--without=")) { applyFeatureSwitches(o, arg.substr(10), false); continue; }
        if (arg == "--shading") { o.hasShadingType = true; o.shadingType = parseShading(readValue("--shading")); continue; }
        if (startsWith(arg, "--shading=")) { o.hasShadingType = true; o.shadingType = parseShading(arg.substr(10)); continue; }
        if (arg == "--texture-filter") { o.hasTextureFilter = true; o.textureFilter = parseTextureFilter(readValue("--texture-filter")); continue; }
        if (startsWith(arg, "--texture-filter=")) { o.hasTextureFilter = true; o.textureFilter = parseTextureFilter(arg.substr(17)); continue; }
        if (arg == "--meridians") { o.hasMeridians = true; o.meridians = std::max(3, std::atoi(readValue("--meridians").c_str())); continue; }
        if (startsWith(arg, "--meridians=")) { o.hasMeridians = true; o.meridians = std::max(3, std::atoi(arg.substr(12).c_str())); continue; }
        if (arg == "--parallels") { o.hasParallels = true; o.parallels = std::max(2, std::atoi(readValue("--parallels").c_str())); continue; }
        if (startsWith(arg, "--parallels=")) { o.hasParallels = true; o.parallels = std::max(2, std::atoi(arg.substr(12).c_str())); continue; }
        if (arg == "--cpu-texture-offset-u") { o.hasCpuTextureOffsetU = true; o.cpuTextureOffsetUTexels = std::atof(readValue("--cpu-texture-offset-u").c_str()); continue; }
        if (startsWith(arg, "--cpu-texture-offset-u=")) { o.hasCpuTextureOffsetU = true; o.cpuTextureOffsetUTexels = std::atof(arg.substr(23).c_str()); continue; }
        if (arg == "--cpu-texture-offset-v") { o.hasCpuTextureOffsetV = true; o.cpuTextureOffsetVTexels = std::atof(readValue("--cpu-texture-offset-v").c_str()); continue; }
        if (startsWith(arg, "--cpu-texture-offset-v=")) { o.hasCpuTextureOffsetV = true; o.cpuTextureOffsetVTexels = std::atof(arg.substr(23).c_str()); continue; }
        if (arg == "--width") { o.width = std::max(1, std::atoi(readValue("--width").c_str())); continue; }
        if (startsWith(arg, "--width=")) { o.width = std::max(1, std::atoi(arg.substr(8).c_str())); continue; }
        if (arg == "--height") { o.height = std::max(1, std::atoi(readValue("--height").c_str())); continue; }
        if (startsWith(arg, "--height=")) { o.height = std::max(1, std::atoi(arg.substr(9).c_str())); continue; }
        if (arg == "--hud") { o.showHud = true; continue; }
        if (startsWith(arg, "--hud=")) { o.showHud = lower(arg.substr(6)) != "off"; continue; }
        throw std::invalid_argument("Unknown option: " + arg);
    }

    if (o.offline && o.offlineOutputPath.empty()) {
        throw std::invalid_argument("--offline requires an output file path");
    }
    return o;
}
